lib/Extension/LanguageServerEvaluatableExpression/Handler/EvaluatableExpressionHandler.php

```diff
@@ -0,0 +1,106 @@
+<?php
+
+namespace Phpactor\Extension\LanguageServerEvaluatableExpression\Handler;
+
+use function Amp\call;
+use Amp\Promise;
+use Microsoft\PhpParser\Parser;
+use Microsoft\PhpParser\Node;
+use Microsoft\PhpParser\Token;
+use Phpactor\Extension\LanguageServerBridge\Converter\PositionConverter;
+use Phpactor\Extension\LanguageServerBridge\Converter\RangeConverter;
+use Phpactor\Extension\LanguageServerEvaluatableExpression\Protocol\EvaluatableExpression;
+use Phpactor\LanguageServerProtocol\Position;
+use Phpactor\LanguageServerProtocol\ServerCapabilities;
+use Phpactor\LanguageServerProtocol\TextDocumentIdentifier;
+use Phpactor\LanguageServer\Core\Handler\CanRegisterCapabilities;
+use Phpactor\LanguageServer\Core\Handler\Handler;
+use Phpactor\LanguageServer\Core\Workspace\Workspace;
+use Phpactor\TextDocument\TextDocumentBuilder;
+use Phpactor\WorseReflection\Core\Util\NodeUtil;
+use Phpactor\TextDocument\ByteOffsetRange;
+use Phpactor\LanguageServerProtocol\Range;
+
+class EvaluatableExpressionHandler implements Handler, CanRegisterCapabilities
+{
+    public function __construct(
+        private Workspace $workspace,
+        private Parser $parser,
+    ) {
+    }
+
+    public function methods(): array
+    {
+        return [
+            'textDocument/xevaluatableExpression' => 'xevaluatableExpression',
+        ];
+    }
+
+    /**
+     * @return Promise<EvaluatableExpression|null>
+     */
+    public function xevaluatableExpression(
+        TextDocumentIdentifier $textDocument,
+        Position $position
+    ): Promise {
+        return call(function () use ($textDocument, $position) {
+            $document = $this->workspace->get($textDocument->uri);
```

Comment:
Here I think PHPStan _would_ understand this if the `function()` was a generator, but it's not.

You can fix this by not using `call()` and returnjing `new Success($result)` instead:
```diff
diff --git a/lib/Extension/LanguageServerEvaluatableExpression/Handler/EvaluatableExpressionHandler.php b/lib/Extension/LanguageServerEvaluatableExpression/Handler/Evaluata>
index f675ef07..df0d8c55 100644
--- a/lib/Extension/LanguageServerEvaluatableExpression/Handler/EvaluatableExpressionHandler.php
+++ b/lib/Extension/LanguageServerEvaluatableExpression/Handler/EvaluatableExpressionHandler.php
@@ -2,6 +2,7 @@

 namespace Phpactor\Extension\LanguageServerEvaluatableExpression\Handler;

+use Amp\Success;
 use function Amp\call;
 use Amp\Promise;
 use Microsoft\PhpParser\Parser;
@@ -43,26 +44,23 @@ class EvaluatableExpressionHandler implements Handler, CanRegisterCapabilities
         TextDocumentIdentifier $textDocument,
         Position $position
     ): Promise {
-        return call(function () use ($textDocument, $position) {
-            $document = $this->workspace->get($textDocument->uri);
-            $offset = PositionConverter::positionToByteOffset($position, $document->text);
-            $document = TextDocumentBuilder::create($document->text)
-                ->uri($document->uri)
-                ->language('php')
-                ->build();
+        $document = $this->workspace->get($textDocument->uri);
+        $offset = PositionConverter::positionToByteOffset($position, $document->text);
+        $document = TextDocumentBuilder::create($document->text)
+            ->uri($document->uri)
+            ->language('php')
+            ->build();

-            $char = substr($document, $offset->toInt(), 1);
+        $char = substr($document, $offset->toInt(), 1);

-            // do not provide evaluatable for whitespace
-            if (trim($char) == '') {
-                return null;
-            }
+        // do not provide evaluatable for whitespace
+        if (trim($char) == '') {
+            return new Success(null);
+        }

-            $rootNode = $this->parser->parseSourceFile((string) $document, $document->uri()?->__toString());
-            $node = $rootNode->getDescendantNodeAtPosition($offset->toInt());
-            $ret = $this->nodeToEvaluatable($node);
-            return $ret;
-        });
+        $rootNode = $this->parser->parseSourceFile((string) $document, $document->uri()?->__toString());
+        $node = $rootNode->getDescendantNodeAtPosition($offset->toInt());
+        return new Success($this->nodeToEvaluatable($node));
     }
```

Reply:
Yes, the async `call()` is a leftover from the `HoverHanlder` that I used as a starting point. But even there, phpstan does not process it correctly, because there is an exception in `phpstan-baseline.neon`

```
		-
			message: "#^Method Phpactor\\\\Extension\\\\LanguageServerHover\\\\Handler\\\\HoverHandler\\:\\:hover\\(\\) should return Amp\\\\Promise\\<Phpactor\\\\LanguageServerProtocol\\\\Hover\\|null\\> but returns Amp\\\\Promise\\<mixed\\>\\.$#"
			count: 1
			path: lib/Extension/LanguageServerHover/Handler/HoverHandler.php
```

I'm not sure about all the architecture, but even though `xevaluatableExpression` does not have any async code in it, leaving it inside `call()` allows the `Amp` loop to process other messages before spending CPU in this call... I could be wrong though...

Reply:
basically the request comes it, it gets routed to here and even if there was some "other" work to do it makes no difference if this is executed now or on the next tick, so I think executing it directly is fine.